sage.py

```diff
@@ -481,1274 +423,920 @@ def getepisodes(action_seq, mcat, plot, debug=False):
+        for epi in episodes:
+            if len(str(epi[2])) == 3:  # If high-severity, then include it
+                mcat = micro[epi[2]].split('.')[1]
+                if obj_only:                         # Experiment: only mcat or mcat + mserv?
+                    vert_name = mcat
+                else:
+                    vert_name = mcat + '|' + epi[4]
+                objectives.add(vert_name)
+    return list(objectives)
+
+
+def _is_severe_sink(vname, sev_sinks):
+    for sink in sev_sinks:
+        if vname.split("|")[-1] == sink:
+            return True
+    return False
+
+
+def _make_vertex_info(episode, in_main_model):
+    start_time = round(episode[0] / 1.0)
+    end_time = round(episode[1] / 1.0)
+    cat = micro[episode[2]].split('.')[1]
+    signs = episode[5]
+    timestamps = episode[6]
+    if episode[3] in in_main_model:  # TODO: this check is useless, since it is always True
+        state_id = '' if len(str(episode[2])) == 1 else '|' + str(episode[3])
+    else:
+        state_id = '|Sink'
+    vert_name = cat + '|' + episode[4] + state_id
+    vert_info = (vert_name, start_time, end_time, signs, timestamps)
+    return vert_info
+
+
+def _get_attack_attempts(state_sequences, victim, objective, in_main_model):
+    team_attack_attempts = dict()
+    observed_obj = set()  # Variants of current objective
+    for att, episodes in state_sequences.items():  # Iterate over (a,v): [episode, episode, episode]
+        if victim != att.split("->")[1]:  # If it's not the right victim, then don't process further
+            continue
+        vertices = []
+        for epi in episodes:
+            vertices.append(_make_vertex_info(epi, in_main_model))
+
+        # If the objective is never reached, don't process further
+        if not sum([True if objective.split("|")[:2] == v[0].split("|")[:2] else False for v in vertices]):
+            continue
+
+        # If it's an episode sequence targeting the requested victim and obtaining the requested objective
+        attempts = []
+        sub_attempt = []
+        for vertex in vertices:  # Cut each attempt until the requested objective
+            sub_attempt.append(vertex)  # Add the vertex in path
+            if objective.split("|")[:2] == vertex[0].split("|")[:2]:  # If it's the objective
+                if len(sub_attempt) <= 1:  # If only a single node, reject
+                    sub_attempt = []
+                    continue
+                attempts.append(sub_attempt)
+                sub_attempt = []
+                observed_obj.add(vertex[0])
+                continue
+        team_attacker = att.split('->')[0]  # Team + attacker
+        if team_attacker not in team_attack_attempts.keys():
+            team_attack_attempts[team_attacker] = []
+        team_attack_attempts[team_attacker].extend(attempts)
+        # team_attack_attempts[team_attacker] = sorted(team_attack_attempts[team_attacker], key=lambda item: item[1])
```

Comment:
Not sure if it needs to be sorted, so keeping it commented as it was

Reply:
idk either